### PR TITLE
feat: gitlab support

### PR DESCRIPTION
**Pull request by noamd-legit**
#### What's being changed?
Adding support for the GitLab platform

#### Check off the following:

- [X] This PR follows the CONTRIBUTION.md guidelines
- [X] I have self-reviewed my changes before submitting the PR

### CODE REVIEW

File: cmd/analyze.go [hunk 1]
```diff
@@ -155,59 +125,20 @@ func executeAnalyzeCommand(cmd *cobra.Command, _args []string) error {
...
-
-	ctx = context_utils.NewContextWithTokenScopes(ctx, githubClient.Scopes())
-
-	opaEngine, err := opa.Load(analyzeArgs.PoliciesPath)
-	if err != nil {
-		return err
+	if analyzeArgs.ScmType == scm_type.GitHub {
+		executor, err = setupGitHub(&analyzeArgs, stdErrLog)
+	} else if analyzeArgs.ScmType == scm_type.GitLab {
+		executor, err = setupGitlab(&analyzeArgs, stdErrLog)
```
Comment (gal-legit) [inline cmd/analyze.go:133]: Gitlab => GitLab *

File: cmd/inject_gitlab.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+func provideGitlabCollectors(ctx context.Context, client *glclient.Client, analyzeArgs *args) []collectors.Collector {
+	var collectorsMapping = map[namespace.Namespace]func(ctx context.Context, client *glclient.Client) collectors.Collector{
+		namespace.Organization: gitlab.NewGroupCollector,
+	}
+
+	var result []collectors.Collector
+	for _, ns := range analyzeArgs.Namespaces {
+		if val, ok := collectorsMapping[ns]; ok {
```
Comment (gal-legit) [inline cmd/inject_gitlab.go:33]: val => creator

File: internal/common/types/types.go [hunk 3]
```diff
@@ -1,6 +1,31 @@
...
+}
+
+func NewRepositoryWithOwner(repositoryWithOwner string, perms permissions.RepositoryRole) RepositoryWithOwner {
+	split := strings.Split(repositoryWithOwner, ",")
+	return RepositoryWithOwner{
+		Owner: split[0],
+		Name:  split[1],
+		Role:  perms,
+	}
+}
```
Comment (gal-legit) [inline internal/common/types/types.go:22]: owner,name,found := strings.Cut() and return error if !found

File: cmd/wire_gen.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
+// Code generated by Wire. DO NOT EDIT.
```
Comment (gal-legit) [inline cmd/wire_gen.go:1]: check if we should really push generated code

File: internal/clients/github/client.go [hunk 5]
```diff
@@ -360,6 +368,137 @@ func (c *client) GetActionsTokenPermissions(url string) (*types.TokenPermissions
...
+		"name":  githubv4.String(repository.Name),
+		"owner": githubv4.String(repository.Owner),
+	}
+
+	err := c.GraphQLClient().Query(c.context, &repo, variables)
+	if err != nil {
+		return false, err
+	}
+
+	return repo.Repository.ViewerPermission == permissions.RepoRoleAdmin, nil
```
Comment (gal-legit) [inline internal/clients/github/client.go:387]: does it work when user is OrgAdmin but not a member of the repo?

File: internal/clients/github/client.go [hunk 6]
```diff
@@ -360,6 +368,137 @@ func (c *client) GetActionsTokenPermissions(url string) (*types.TokenPermissions
...
+}
+
+func unique(slice []commontypes.RepositoryWithOwner) []commontypes.RepositoryWithOwner {
+	keys := make(map[string]bool)
+	var list []commontypes.RepositoryWithOwner
+	for _, entry := range slice {
+		key := entry.String()
+		if _, value := keys[key]; !value {
```
Comment (gal-legit) [inline internal/clients/github/client.go:395]: value => found

File: internal/clients/github/client.go [hunk 7]
```diff
@@ -360,6 +368,137 @@ func (c *client) GetActionsTokenPermissions(url string) (*types.TokenPermissions
...
+
+	err := c.GraphQLClient().Query(c.context, &repo, variables)
+	if err != nil {
+		return false, err
+	}
+
+	return repo.Repository.ViewerPermission == permissions.RepoRoleAdmin, nil
+}
+
+func unique(slice []commontypes.RepositoryWithOwner) []commontypes.RepositoryWithOwner {
```
Comment (gal-legit) [inline internal/clients/github/client.go:390]: uniqueRepos

File: internal/clients/gitlab/client.go [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	git, err := gitlab.NewClient(token, config...)
+	if err != nil {
+		return nil, err
+	}
+
+	// all groups
+	if len(orgs) == 0 {
+		orgs = []string{""}
```
Comment (gal-legit) [inline internal/clients/gitlab/client.go:39]: const allSomething = ""

File: internal/clients/gitlab/client.go [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func (c *Client) Repositories() ([]types.RepositoryWithOwner, error) {
+	var result []types.RepositoryWithOwner
+
+	dummy := gitlab.MaintainerPermissions
```
Comment (gal-legit) [inline internal/clients/gitlab/client.go:95]: maintainerPerms := gitlab.MaintainerPermissions

File: internal/clients/gitlab/client.go [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+func (c *Client) Groups() ([]*gitlab.Group, error) {
+	if groups, found := c.cache.Get(orgsCacheKeys); found {
+		return groups.([]*gitlab.Group), nil
+	}
+
+	var result []*gitlab.Group
+
+	dummy := true
```
Comment (gal-legit) [inline internal/clients/gitlab/client.go:138]: dummy -> ownedGroups

File: internal/collectors/gitlab/collection_context.go [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+func newCollectionContext(group *gitlab.Group, roles []permissions.Role) collectionContext {
+	return collectionContext{
+		group: group,
+		roles: roles,
+	}
+}
+
+func (c collectionContext) IsEnterprise() bool {
```
Comment (gal-legit) [inline internal/collectors/gitlab/collection_context.go:20]: IsEnterprise() => IsPremium()

Comment (gal-legit) [reply internal/collectors/gitlab/collection_context.go:20]: and add a note about GitLab (currently no policy requires premium)

File: internal/common/types/types.go (hunk 3 above)
Comment (gal-legit) [reply internal/common/types/types.go:22]: and probably , => /

File: internal/opa/opa.go [hunk 12]
```diff
@@ -49,8 +50,12 @@ func Load(policyPaths []string) (opa_engine.Enginer, error) {
 	return engine, nil
 }
 
-func loadModules() ([]*ast.Module, error) {
-	return loadModulesFromFs(policies.Bundle, path.Dir(""))
+func loadModules(scmType scm_type.ScmType) ([]*ast.Module, error) {
+	if scmType == scm_type.GitHub {
```
Comment (gal-legit) [inline internal/opa/opa.go:54]: switch {
  github
  gitlab
  default: err
 }

File: internal/clients/github/client.go (hunk 5 above)
Comment (noamd-legit) [reply internal/clients/github/client.go:387]: yes

**Review by gal-legit — DISMISSED** [review]

**Review by gal-legit — APPROVED** [review]

**Pull request merged**